compiler/rustc_typeck/src/check/fn_ctxt/suggestions.rs

```diff
@@ -491,12 +495,23 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 | ExprKind::Block(..)
                     if expression.can_have_side_effects() =>
                 {
-                    err.span_suggestion(
-                        expression.span.shrink_to_hi(),
-                        "consider using a semicolon here",
-                        ";".to_string(),
-                        Applicability::MachineApplicable,
-                    );
+                    if needs_block {
+                        err.multipart_suggestion(
+                            "consider using a semicolon here",
```

Comment:
I'm not sure about the wording: should we mention the block, too?

Reply:
Nah, the suggestion is obvious, and the semicolon is the important part, so mentioning that is sufficient imo

Reply:
Nah, the suggestion is obvious, and the semicolon is the important part, so mentioning that is sufficient imo